### PR TITLE
PE-42104 - Adding Jobstat to SAT CLI (continuation of PR-38)

### PR DESCRIPTION
**Pull request by vinaybabug-hpe**
This commit implements `sat jobstat`. This command gets a system-wide view of
  applications and displays it in a pretty table format. This commit also adds
  a change log entry and man page for the command.

### CODE REVIEW

File: docs/man/sat-jobstat.8.rst
Comment (ekamin-hpe) [inline docs/man/sat-jobstat.8.rst:7]: Match the number of characters in the underline/overline to the title.

```suggestion
-------------------------------------------------------
Jobstat - Check status of job(s) and its application(s)
-------------------------------------------------------
```

**Review by ekamin-hpe — CHANGES_REQUESTED** [review]
There are still some unaddressed comments on the original PR:

https://github.com/Cray-HPE/sat/pull/38#discussion_r950511118
https://github.com/Cray-HPE/sat/pull/38#discussion_r950513754
https://github.com/Cray-HPE/sat/pull/38#discussion_r950516301
https://github.com/Cray-HPE/sat/pull/38#discussion_r950517035
https://github.com/Cray-HPE/sat/pull/38#discussion_r950519334
https://github.com/Cray-HPE/sat/pull/38#discussion_r950525281
https://github.com/Cray-HPE/sat/pull/38#discussion_r950525875
https://github.com/Cray-HPE/sat/pull/38#discussion_r950526388

I am also still curious if there is going to be authentication on the API endpoint as I mentioned in this comment, and we should be using the config file field for the API gateway hostname rather than hard-coding it: https://github.com/Cray-HPE/sat/pull/38#discussion_r968926035

Finally, as I brought up in that last comment, I don't really like that we are using `requests` in a different way than the other commands, which all use the `APIGatewayClient` class by deriving a subclass from it. Maybe that is something we can help with, if my last comment was not clear on how to proceed.

File: docs/man/sat-jobstat.8.rst
Comment (ekamin-hpe) [inline docs/man/sat-jobstat.8.rst:21]: Sorry for not noticing this earlier, but could this line be wrapped at 80 characters?

File: sat/cli/jobstat/parser.py
Comment (ekamin-hpe) [inline sat/cli/jobstat/parser.py:44]: nit: "relevent" is misspelled.

File: sat/cli/jobstat/main.py
Comment (ekamin-hpe) [inline sat/cli/jobstat/main.py:38]: I think this would potentially work though it won't work if the API Gateway is authenticated.

We don't usually instantiate `APIGatewayClient` directly, instead we always create subclasses for the different services.

Here is a gist explaining what I think should change: https://gist.github.hpe.com/eli-kamin/3675a843e0b343f3360f2bcd11327da1

File: sat/cli/jobstat/main.py
Comment (ekamin-hpe) [inline sat/cli/jobstat/main.py:70]: I would prefer that we don't catch the bare exception class here.

We should most likely only need to catch HTTP errors coming from the API. This comment is also addressed in the gist I left: https://gist.github.hpe.com/eli-kamin/3675a843e0b343f3360f2bcd11327da1

File: docs/man/sat-jobstat.8.rst
Comment (ekamin-hpe) [inline docs/man/sat-jobstat.8.rst:32]: Please add an empty line between lines 31 and 32. As it currently exists, compiling the man pages gives a warning:
```
sat-jobstat.8.rst:32: (WARNING/2) Definition list ends without a blank line; unexpected unindent.
```

**Review by ekamin-hpe — CHANGES_REQUESTED** [review]
This is getting much closer, though I would really still like to refactor the API client code to be consistent with the other commands in SAT. I hope the gist I provided helps with that.

Additionally, some level of testing would be necessary - optimally, we would include a description of testing performed in the commit message. Is this status checker a service that will be available generally on internal systems? Is it part of some product stream?

Once we're ready to move forward with this, you'll need to squash all the commits in this pull request to one so we can keep the commit history clean, and I will help with getting it checked in.

File: sat/cli/jobstat/main.py
Comment (ekamin-hpe) [inline sat/cli/jobstat/main.py:45]: Can this case be removed again? I think in my gist I did not remove it, sorry about that.

File: sat/apiclient/state_checker.py
Comment (ekamin-hpe) [inline sat/apiclient/state_checker.py:33]: Can you please add an extra blank line here? Otherwise we get a code style issue:
```
mbp-ekamin:sat ekamin$ pycodestyle --config pycodestyle.conf sat tests
sat/apiclient/state_checker.py:34:1: E302 expected 2 blank lines, found 1
```

File: sat/apiclient/state_checker.py
Comment (ekamin-hpe) [inline sat/apiclient/state_checker.py:25]: Sorry that I didn't do this in my original gist, but I am thinking it might be a good idea to rename the class `JobstatStateChecker` instead of just "state checker", otherwise it is really vague. There are so many different clients in SAT, that it might be misleading and sound like this is for HSM state or some other "state" service.

That would entail this in a few different places, specifically:
* The name of the class
* The name of this file
* Any mentions in docs or error messages (e.g. lines 38 and 48)

**Review by ekamin-hpe — COMMENTED** [review]
A few more comments, but this is looking very good. Prior to merging I would like the branch to be rebased and have the various commits squashed into one, and if you could put a test description in the commit message that would be helpful. I am also thinking unit tests could still be helpful, though perhaps not urgently needed if manual testing can be done.

I am also going to change the target of this PR to be `feature/PE-42104-sat-jobstat`. What will need to happen once this is ready to be checked in is that we will merge it to that target branch, at which point we will verify the build is working and potentially add any last-minute fixes. We might add unit tests at this stage as well. Then we will check into integration.

File: sat/cli/jobstat/main.py
Comment (ekamin-hpe) [inline sat/cli/jobstat/main.py:46]: I believe we now need to account for the case when `application_data` is an empty list, otherwise I see the following error:
```
mbp-ekamin:sat ekamin$ sat jobstat
...
Traceback (most recent call last):
  File "/Users/ekamin/.virtualenvs/sat/bin/sat", line 33, in <module>
    sys.exit(load_entry_point('sat', 'console_scripts', 'sat')())
  File "/Users/ekamin/code/sat/sat/main.py", line 120, in main
    subcommand(args)
  File "/Users/ekamin/code/sat/sat/cli/jobstat/main.py", line 47, in do_jobstat
    tuple(application_data[0].keys()),
IndexError: list index out of range
```

File: sat/apiclient/state_checker.py
Comment (ekamin-hpe) [inline sat/apiclient/state_checker.py:50]: I am getting some weird behavior running through this code. The state checker service seems to be returning `200` code responses but are actually error cases, and as a result `APIError` is not being raised here as we would expect.

```
>>> resp = self.get('all')
>>> resp
<Response [200]>
```
But in the actual JSON there is an error:
```
>>> resp.json()
{'errorcode': 4, 'errormessage': 'Problem processing connection request', 'timereported': '2022-09-19T18:34:24', 'jobstat': []}
```

It seems like this should really be addressed in the `jobstat` service (codes of 4XX or 5XX will automatically raise an error). Maybe you could work around it here by raising an `APIError` if the returned JSON contained an `errorcode` field, but I wouldn't say that's an optimal solution.

```
        try:
            response = self.get('all').json()
            error_code = response.get('errorcode')
            if error_code is not None and error_code != 0:
                raise APIError(f'{response.get("errormessage", "error in state checker API")}')
            return response['jobstat']
        except APIError as err:
            raise APIError(f'{err_prefix}: {err}')
```

File: sat/cli/jobstat/main.py
Comment (ekamin-hpe) [inline sat/cli/jobstat/main.py:46]: I am a little confused by the most recent change on this line, but I think you probably meant to do this as `is not None` rather than `== None`, otherwise you are only creating the report if the response looks like this:
```
{"jobstat": null}
```
(i.e. it contained a `'jobstat'` key and the value was `NoneType`).

With this change, I also don't think it's going to be correctly handling the case when the response is missing a `jobstat` key, the command will fail with an uncaught `KeyError` coming from this line.

Assuming you did mean to say `if application_data['jobstat'] is not None`, then If the response does contain a `jobstat` key but returns an empty list, I think you will get an uncaught `IndexError` on line 48 trying to get the 0th element of an empty list.

I think it would be good to also account for the case when the response contains the `jobstat` key but a `null` value (i.e. that case I show at the top of this comment), once again I think that would result in an unhandled exception coming from line 48.

Just in general I think it would still be cleaner to be returning just the value pointed to by the `"jobstat"` key - this version makes the `except KeyError` on line 55 of `state_checker.py` no longer necessary - and I don't think we are correctly handling that case now.

For me I think I would still prefer something along these lines, though there might be cases that I am not thinking of: https://gist.github.hpe.com/eli-kamin/aa8f60f88dd465507f6410ee568e98ba

File: sat/cli/jobstat/main.py
Comment (ekamin-hpe) [inline sat/cli/jobstat/main.py:63]: I would much prefer if we could keep this error handling in the `StateCheckerClient` class in the form of raising an `APIError` (see other comment), but if we must do it here, let's use `LOGGER.error` instead of `print`, so that the error goes to `stderr`.

**Review by ekamin-hpe — CHANGES_REQUESTED** [review]
I have some comments regarding error handling.

File: sat/cli/jobstat/main.py
Comment (ekamin-hpe) [reply sat/cli/jobstat/main.py:46]: Are we attempting to handle the case when the API returns both valid jobstat results as well as an error message? It seems like maybe you are trying to do that, not sure.

Comment (vinaybabug-hpe) [reply sat/cli/jobstat/main.py]: Our response always has jobstat e.g. 
`{'machine': 'slice', 'errorcode': 2, 'errormessage': 'Unable to query Elasticsearch', 'timereported': '2022-09-20T08:26:02', 'jobstat': []}`
We have machine just in case for multi tenancy.  A non-zero errorcode indicates failure so we use that instead of suggested checks.

Comment (vinaybabug-hpe) [reply sat/cli/jobstat/main.py]: Response structure is always the same from the server, just values change.

File: sat/apiclient/JobstatStateChecker.py
Comment (ekamin-hpe) [inline sat/apiclient/JobstatStateChecker.py:53]: This is pretty close to what I was hoping for, thanks for addressing all the feedback here.

I see you mentioned that the response objects always have `errorcode` and `errormessage` keys (even in the successful case) but I think it would be good to be more defensive and handle when they aren't there, i.e. first check that the expected keys exist.

Otherwise, in the case when there is successful `jobstat` data but there is no errorcode, we will fail with an `APIError` (due to missing key). Other (edge cases) thoughts I have are how we are handling if `errorcode` is a null type, and how we are handling if the 'errormessage' key is missing.

Just in case the behavior of the API changes, or something unexpected happens.

How would you feel about this approach, I think it's a bit more defensive:
```suggestion
            if 'errorcode' in response and response['errorcode']:
                raise APIError(response.get('errormessage', 'unknown error'))
```

**Review by ekamin-hpe — COMMENTED** [review]
I have one final comment, I think after that I would be ready to approve. There seems to also be a conflict in the change log that needs resolving.

When you're done pushing changes, could you please squash all your commits into one and include a short description of the testing performed? I personally was unable to test this command in the successful case, every time I tried running it I only got errors from the service.

File: sat/cli/jobstat/main.py
Comment (jack-stanek-hpe) [inline sat/cli/jobstat/main.py:54]: FWIW, these arguments aren't needed. When the `no_headings` and `no_borders` arguments are `None` (the default), the `Report` class looks up the config file values. See here: https://github.com/vinaybabug-hpe/sat/blob/993cf683b89798958244ac2ab4473f5e8650bef0/sat/report.py#L102-L114

**Review by jack-stanek-hpe — COMMENTED** [review]
Where can I find a specification for the jobstat API (e.g. OpenAPI or Swagger specs?)

File: sat/apiclient/JobstatStateChecker.py
Comment (vinaybabug-hpe) [reply sat/apiclient/JobstatStateChecker.py:35]: For changes regarding APIError; I will defer to sat team as the current codebase does not mention changes.

File: sat/cli/jobstat/main.py
Comment (vinaybabug-hpe) [reply sat/cli/jobstat/main.py:54]: removed them

File: sat/apiclient/jobstat.py
Comment (ekamin-hpe) [inline sat/apiclient/jobstat.py:53]: I would slightly prefer `APIError(response.get('errormessage'))` so that we don't raise a `KeyError` if `errormessage` is unexpectedly missing. Possibly, `APIError(response.get('errormessage', 'unknown error'))`.

File: sat/apiclient/jobstat.py
Comment (ekamin-hpe) [inline sat/apiclient/jobstat.py:54]: I would actually say we don't necessarily need to check `if 'jobstat' in response` here, because this would be handled by the `except KeyError` on line 61. If the `jobstat` key is missing and we didn't get any error reported, I think we should be raising an `APIError`.

File: sat/apiclient/jobstat.py
Comment (ekamin-hpe) [inline sat/apiclient/jobstat.py:56]: This line is not necessary because functions will return `None` unless otherwise specified. If we want a default case, I'd make it an empty list, but I would prefer if the `jobstat` key is missing to be raising an error (as I mentioned in the comment above).

The reason I am thinking of the `errorcode` and `jobstat` fields as different in terms of how we should be checking them is that in the case that the `errorcode` field is missing but `jobstat` key is not, I think we should be able to return the jobstat data without an error. But in the cases:
* when both the `jobstat` key and the `errorcode` key are missing
* when the `errorcode` was 0 but the `jobstat` key was still missing

These are cases when I think we should be raising an error.

File: docs/man/sat-jobstat.8.rst
Comment (ekamin-hpe) [inline docs/man/sat-jobstat.8.rst:5]: nit: now that line 6 has been shortened could you shorten these lines to match as well? (This line and line 7).

**Review by ekamin-hpe — COMMENTED** [review]
Sorry to keep nitpicking the implementation of `JobstatClient.get_all`, I appreciate your continuing to refine it. I have a few more comments about this version. Here is what I was thinking for this function:

https://gist.github.hpe.com/eli-kamin/59d01cad892f88e5a73bc2ce26ab68ed

File: sat/apiclient/jobstat.py
Comment (vinaybabug-hpe) [reply sat/apiclient/jobstat.py:54]: well, missing errorcode/errormessage would have resulted in the same. This just handles the case more gracefully...

File: sat/apiclient/jobstat.py
Comment (vinaybabug-hpe) [reply sat/apiclient/jobstat.py:56]: I don't see this happening with the current server implementation. Only memory corruption could result in such scenarios.

**Comment by vinaybabug-hpe** [comment]
I am not able to sync with the main sat codebase as this code is a fork from Elmir's code/fork. Can we take this to a branch in Cray-HPE/sat (not master)?

File: sat/apiclient/jobstat.py
Comment (ekamin-hpe) [reply sat/apiclient/jobstat.py]: My worry is that the server implementation could change. I don't think there's ever really a good case for returning `None` from this function, I think it should either return valid results or raise an error.

File: sat/apiclient/jobstat.py
Comment (ekamin-hpe) [reply sat/apiclient/jobstat.py]: I was thinking of the case when the server returned valid jobstat results but for some reason didn't return errorcode/errormessage keys. I think in that case we would not want to raise an error.

File: sat/apiclient/JobstatStateChecker.py
Comment (ekamin-hpe) [reply sat/apiclient/JobstatStateChecker.py]: Depending on the timing, this may be something that the SAT team will need to address after this has been merged.

Comment (jack-stanek-hpe) [reply sat/apiclient/JobstatStateChecker.py]: Correct, no changes are needed at this point.

**Review by ekamin-hpe — COMMENTED** [review]
Just a quick comment on the commit message, "Refactored code according to comments by using PIGatewayClient" is missing a letter, should be "Refactored code according to comments by using APIGatewayClient"

**Review by ekamin-hpe — APPROVED** [review]
Looks good, thank you again for the sustained effort to address all the feedback. This will be merged to the `feature/PE-42104-sat-jobstat` branch in the main Cray-HPE/sat repository, at which point we will take it over and ensure it passes all the necessary checks to go into integration.

**Pull request merged**
